desc/objectives/nae_utils.py

```diff
@@ -0,0 +1,239 @@
+        stellarator symmetric for R i.e. R(-theta,-phi) = R(theta,phi)
+        and Z_1_1_n uses the Zbasis_sin as the term is cos(theta)*sin(phi)
+        since Z(-theta,-phi) = - Z(theta,phi) for Z stellarator symmetry.
+    """
+    phi = qsc.phi
+
+    R0 = qsc.R0_func(phi)
+    dR0_dphi = qsc.R0p
+    dZ0_dphi = qsc.Z0p
+    # normal and binormal vector components
+    # Spline interpolants for the cylindrical components of the Frenet-Serret frame:
+    # these are functions of phi (toroidal cylindrical angle)
+    k_dot_R = qsc.normal_R_spline(phi)
+    k_dot_phi = qsc.normal_phi_spline(phi)
+    k_dot_Z = qsc.normal_z_spline(phi)
+    tau_dot_R = qsc.binormal_R_spline(phi)
+    tau_dot_phi = qsc.binormal_phi_spline(phi)
+    tau_dot_Z = qsc.binormal_z_spline(phi)
+
+    # use untwisted, which accounts for when NAE has QH symmetry,
+    # and the poloidal angle is a helical angle.
+    # we want the untwisted angle
+    X1c = qsc.X1c_untwisted
+    X1s = qsc.X1s_untwisted
+    Y1c = qsc.Y1c_untwisted
+    Y1s = qsc.Y1s_untwisted
+
+    R_1_1 = X1c * (k_dot_R - k_dot_phi * dR0_dphi / R0) + Y1c * (
+        tau_dot_R - tau_dot_phi * dR0_dphi / R0
+    )
+    R_1_neg1 = Y1s * (tau_dot_R - tau_dot_phi * dR0_dphi / R0) + X1s * (
+        k_dot_R - k_dot_phi * dR0_dphi / R0
+    )
+
+    Z_1_1 = X1c * (k_dot_Z - k_dot_phi * dZ0_dphi / R0) + Y1c * (
+        tau_dot_Z - tau_dot_phi * dZ0_dphi / R0
+    )
+    Z_1_neg1 = Y1s * (tau_dot_Z - tau_dot_phi * dZ0_dphi / R0) + X1s * (
+        k_dot_Z - k_dot_phi * dZ0_dphi / R0
+    )
+
+    nfp = qsc.nfp
+    if desc_eq.sym:
+        Rbasis = FourierSeries(N=desc_eq.N, NFP=nfp, sym="cos")
+        Zbasis = FourierSeries(N=desc_eq.N, NFP=nfp, sym="cos")
+        Rbasis_sin = FourierSeries(N=desc_eq.N, NFP=nfp, sym="sin")
+        Zbasis_sin = FourierSeries(N=desc_eq.N, NFP=nfp, sym="sin")
+    else:
+        Rbasis = FourierSeries(N=desc_eq.N, NFP=nfp, sym=False)
+        Zbasis = FourierSeries(N=desc_eq.N, NFP=nfp, sym=False)
+        Rbasis_sin = FourierSeries(N=desc_eq.N, NFP=nfp, sym=False)
+        Zbasis_sin = FourierSeries(N=desc_eq.N, NFP=nfp, sym=False)
+
+    grid = LinearGrid(M=0, L=0, zeta=phi, NFP=nfp)
+    Rtrans = Transform(grid, Rbasis, build_pinv=True, method="auto")
+    Ztrans = Transform(grid, Zbasis, build_pinv=True, method="auto")
+    Rtrans_sin = Transform(grid, Rbasis_sin, build_pinv=True, method="auto")
+    Ztrans_sin = Transform(grid, Zbasis_sin, build_pinv=True, method="auto")
+
+    R_1_1_n = Rtrans.fit(R_1_1)
+    R_1_neg1_n = Rtrans_sin.fit(R_1_neg1)
+
+    Z_1_1_n = Ztrans_sin.fit(Z_1_1)
+    Z_1_neg1_n = Ztrans.fit(Z_1_neg1)
+
+    bases = {}
+    bases["Rbasis_cos"] = Rbasis
+    bases["Rbasis_sin"] = Rbasis_sin
+    bases["Zbasis_cos"] = Zbasis
+    bases["Zbasis_sin"] = Zbasis_sin
+
+    coeffs = {}
+    coeffs["R_1_1_n"] = R_1_1_n
+    coeffs["R_1_neg1_n"] = R_1_neg1_n
+    coeffs["Z_1_1_n"] = Z_1_1_n
+    coeffs["Z_1_neg1_n"] = Z_1_neg1_n
```

Comment:
maybe we could rename these `"R_1_+1_n"` and `"R_1_-1_n"`?

Reply:
With `neg1` I can make the dictionary keys line up better with the variable names I have used, is my defense for keeping it